File: doric-android/doric/src/main/java/pub/doric/performance/DoricPerformanceProfile.java
/*
 * Copyright [2019] [Doric.Pub]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package pub.doric.performance;

import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import pub.doric.DoricRegistry;

/**
 * @Description: pub.doric.performance
 * @Author: pengfei.zhou
 * @CreateDate: 3/29/21
 */
public class DoricPerformanceProfile {
    private static final String TAG = DoricPerformanceProfile.class.getSimpleName();
    public static final String PART_INIT = "Init";
    public static final String STEP_CREATE = "Create";
    public static final String STEP_Call = "Call";
    public static final String STEP_DESTROY = "Destroy";
    public static final String STEP_RENDER = "Render";
    private static final String MARK_PREPARE = "prepare";
    private static final String MARK_START = "start";
    private static final String MARK_END = "end";
    private final String name;

    private boolean enable = DoricRegistry.isEnablePerformance();
    private static final Handler performanceHandler;
    private final Set<AnchorHook> hooks = new HashSet<>();

    static {
        HandlerThread performanceThread = new HandlerThread("DoricPerformance");
        performanceThread.start();
        performanceHandler = new Handler(performanceThread.getLooper());
    }

    public interface AnchorHook {
        void onAnchor(String name, long prepare, long start, long end);
    }

    public interface GlobalAnchorHook extends AnchorHook {
        void onAnchor(DoricPerformanceProfile profile, String name, long prepare, long start, long end);
    }

    public DoricPerformanceProfile(String name) {
        this.name = name;
    }

    public String getName() {
        return this.name;
    }

    public void addAnchorHook(AnchorHook hook) {
        this.hooks.add(hook);
    }

    public void removeAnchorHook(AnchorHook hook) {
        this.hooks.remove(hook);
    }

    public void enable(boolean enable) {
        this.enable = enable;
    }

    private final Map<String, Long> anchorMap = new HashMap<>();

    private void markAnchor(final String eventName) {
        if (!enable) {
            return;
        }
        performanceHandler.post(new Runnable() {
            @Override
            public void run() {
                long time = System.currentTimeMillis();
                anchorMap.put(eventName, time);
            }
        });
    }

    private String getPrepareAnchor(String anchorName) {
        return anchorName + "#" + MARK_PREPARE;
    }

    private String getStartAnchor(String anchorName) {
        return anchorName + "#" + MARK_START;
    }

    private String getEndAnchor(String anchorName) {
        return anchorName + "#" + MARK_END;
    }

    public void prepare(String anchorName) {
        markAnchor(getPrepareAnchor(anchorName));
    }

    public void start(String anchorName) {
        markAnchor(getStartAnchor(anchorName));
    }

    public void end(String anchorName) {
        markAnchor(getEndAnchor(anchorName));
        print(anchorName);
    }

    private void print(final String anchorName) {
        if (!enable) {
            return;
        }
        performanceHandler.post(new Runnable() {
            @Override
            public void run() {
                Long prepare = anchorMap.get(getPrepareAnchor(anchorName));
                Long start = anchorMap.get(getStartAnchor(anchorName));
                Long end = anchorMap.get(getEndAnchor(anchorName));
                if (end == null) {
                    end = System.currentTimeMillis();
                }
                if (start == null) {
                    start = end;
                }
                if (prepare == null) {
                    prepare = start;
                }
                for (AnchorHook hook : hooks) {
                    if (hook instanceof GlobalAnchorHook) {
                        ((GlobalAnchorHook) hook).onAnchor(DoricPerformanceProfile.this, anchorName, prepare, start, end);
                    } else {
                        hook.onAnchor(anchorName, prepare, start, end);
                    }
                }
            }
        });
    }
}

File: doric-iOS/Pod/Classes/Performance/DoricPerformanceProfile.m
/*
 * Copyright [2019] [Doric.Pub]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//
//  DoricPerformanceProfile.m
//  DoricCore
//
//  Created by pengfei.zhou on 2021/3/29.
//  

#import "DoricPerformanceProfile.h"
#import "DoricRegistry.h"
#import "DoricSingleton.h"

@interface DoricPerformanceProfile ()
@property(nonatomic, strong) dispatch_queue_t anchorQueue;
@property(nonatomic, assign) BOOL enable;
@property(nonatomic, strong) NSHashTable<id <DoricPerformanceAnchorHookProtocol>> *hooks;
@property(nonatomic, strong) NSMutableDictionary <NSString *, NSNumber *> *anchorMap;
@end

@implementation DoricPerformanceProfile
- (instancetype)initWithName:(NSString *)name {
    if (self = [super init]) {
        _name = name;
        _anchorQueue = dispatch_queue_create("doric.performance.profile", DISPATCH_QUEUE_SERIAL);
        _anchorMap = [NSMutableDictionary new];
        _enable = DoricSingleton.instance.enablePerformance;
        _hooks = [NSHashTable weakObjectsHashTable];
    }
    return self;
}

- (void)addAnchorHook:(id)hook {
    __weak typeof(self) _self = self;
    dispatch_async(self.anchorQueue, ^{
        __strong typeof(_self) self = _self;
        if (!self) return;
        
        [self.hooks addObject:hook];
    });
}

- (void)removeAnchorHook:(id)hook {
    __weak typeof(self) _self = self;
    dispatch_async(self.anchorQueue, ^{
        __strong typeof(_self) self = _self;
        if (!self) return;
        
        [self.hooks removeObject:hook];
    });
}

- (void)enable:(bool)enable {
    self.enable = enable;
}

- (NSString *)getPrepareAnchor:(NSString *)anchorName {
    return [NSString stringWithFormat:@"%@#prepare", anchorName];
}

- (NSString *)getStartAnchor:(NSString *)anchorName {
    return [NSString stringWithFormat:@"%@#start", anchorName];
}

- (NSString *)getEndAnchor:(NSString *)anchorName {
    return [NSString stringWithFormat:@"%@#end", anchorName];
}

- (void)markAnchor:(NSString *)eventName {
    if (!self.enable) {
        return;
    }
    NSUInteger time = (NSUInteger) ([[NSDate date] timeIntervalSince1970] * 1000);
    dispatch_async(self.anchorQueue, ^{
        self.anchorMap[eventName] = @(time);
    });
}


- (void)prepare:(NSString *)anchorName {
    [self markAnchor:[self getPrepareAnchor:anchorName]];
}

- (void)start:(NSString *)anchorName {
    [self markAnchor:[self getStartAnchor:anchorName]];
}

- (void)end:(NSString *)anchorName {
    [self markAnchor:[self getEndAnchor:anchorName]];
    [self print:anchorName];
}

- (void)print:(NSString *)anchorName {
    if (!self.enable) {
        return;
    }
    __weak typeof(self) _self = self;
    dispatch_async(self.anchorQueue, ^{
        __strong typeof(_self) self = _self;
        if (!self) return;
        
        NSNumber *prepare = self.anchorMap[[self getPrepareAnchor:anchorName]];
        NSNumber *start = self.anchorMap[[self getStartAnchor:anchorName]];
        NSNumber *end = self.anchorMap[[self getEndAnchor:anchorName]];
        if (!end) {
            end = @( (NSUInteger) [[NSDate date] timeIntervalSince1970] * 1000);
        }
        if (!start) {
            start = end;
        }
        if (!prepare) {
            prepare = start;
        }
        for (id <DoricPerformanceAnchorHookProtocol> hook in self.hooks) {
            if ([hook conformsToProtocol:@protocol(DoricPerformanceGlobalAnchorHookProtocol)]) {
                [(id <DoricPerformanceGlobalAnchorHookProtocol>) hook onAnchorName:anchorName prepare:prepare start:start end:end in:self];
            } else {
                [hook onAnchorName:anchorName prepare:prepare start:start end:end];
            }
        }
    });
}
@end
